### DISCUSSION TITLE
Issue compatibility with Auth0

### DISCUSSION THREAD

**Discussion by williamarmilleiCW** [spatie/laravel-permission#2343] [Q&A] [answered]
Hi,

I have been trying to integrate the latest version of Auth0 with spatie/laravel-permission ‘and Spatie\Activitylog’ in my Laravel 9 application; it was working fine with Laravel 8, but the latest Auth0 version changes drastically auth.php file, changing the guards used by Spatie plugins. I believe this is causing quite a few conflicts. I might be wrong with my diagnoses, probably the reasons could be different, but the following is the error which appears when trying to login into my application:

On “Spatie/permissions”:
Spatie\Permission\PermissionRegistrar::Spatie\Permission{closure}(): Argument #1 ($user) must be of type Illuminate\Contracts\Auth\Access\Authorizable, Auth0\Laravel\Model\Stateful\User given, called in /var/www/html/vendor/laravel/framework/src/Illuminate/Auth/Access/Gate.php on line 553

On “Spatie\Activitylog”:
Spatie\Activitylog\CauserResolver::getDefaultCauser(): Return value must be of type ?Illuminate\Database\Eloquent\Model, Auth0\Laravel\Model\Stateful\User returned

As you see above in each instance Spatie expects an Authorizable instead an Model\Stateful\User is returned by Auth0, so, triggering the error.

Could you please help me to understand how to solve the issue?
Thanks,
William

**Comment by parallels999** [comment]
Difficult to answer without a [minimal demo app](https://spatie.be/docs/laravel-permission/v5/basic-usage/new-app) on a fresh laravel installation

**Reply by parallels999** [reply]
Did you try without that?
Try to do a PR to fix it, just remove the `Authorizable` contract on datatype arg

**Reply by williamarmilleiCW** [reply]
It would be great if you have both Auth0 and spatie-permissions  installed  in a Laravel 9 demo and test it. Removing the contract Authorizable could solve the issue with that specific function registerPermissions but I am afraid in other part of laravel-permission spatie code might have conflict with Auth0.

**Reply by parallels999** [reply]
Of course, it is possible, but when you send a PR you have to test other parts of laravel-permission too, since you are the only one performing this implementation, only you could foresee what problem you would have

**Reply by williamarmilleiCW** [reply]
To be honest I don't know if I should be the one to fix it, I thought in here I should report the problem and those who have developed the code would fix it. Am I wrong? So, the Spatie team, if they are interested in having laravel-permission working together with Auth0 latest version in a Laravel 9 could have this settings in place in a demo and see which errors come up. Or I am wrong? Should I be the one to fix it?

**Reply by erikn69** [reply]
laravel-permission was made to work with Authorizable, any bug in that environment is reviewed by spatie team, but it is not guaranteed to work with other implementations, the PRs of the Auth0 community will be reviewed and accepted if they do not affect the original functionality
Spatie team or contributors can't test Auth0, so
>Should I be the one to fix it?

If you really need it, **yes**, maybe you can look for an already working alternative

**Comment by williamarmilleiCW** [comment]
Basically, go to the Spatie PermissionRegistrar.php function, at line 86 you have registerPermission() function, the $user model passed has to be of the type "Authorizable", but the one passed when Auth0 is installed is "Auth0\Laravel\Model\Stateful\User" causing the issue.  Basically Auth0 changes the model user type, from Authorizable to Auth0 one, your code does not recognize it and trigger the error. Is this more clear?


![image](https://user-images.githubusercontent.com/121885095/222448887-95c7a248-e741-4b45-8173-7c0b6c59d7a1.png)

**Comment by angeljqv** [comment]
> latest Auth0 version changes drastically auth.php file, changing the guards used by Spatie plugins

Examples??

**Reply by angeljqv** [reply]
What about a custom UserProvider like this https://github.com/auth0/laravel-auth0/pull/264, just add Authorizable trait, and with that it should work

**Comment by parallels999** [comment, marked as answer]
Already answered here: https://github.com/auth0/laravel-auth0/issues/350#issuecomment-1455783619

**Reply by williamarmilleiCW** [reply]
Thanks @parallels999, as you might notice that's my comment  :-) At the end I have found the solution

**Reply by parallels999** [reply]
Yes, i notice, leaving the link to the solution will help others with the same problem in the future

**Reply by williamarmilleiCW** [reply]
Sorry about that.
